### ISSUE TITLE
Shared and mutated state between training and validation callbacks

### ISSUE THREAD

**Issue by IvanUkhov** [tensorflow/tensorflow#32819]
**System information**
- Have I written custom code: Yes
- OS Platform and Distribution: Datalab
- Mobile device: NA
- TensorFlow installed from: Binary
- TensorFlow version: 1.15.0rc1
- Python version: 3.5.6
- Bazel version: NA
- GCC/Compiler version: NA
- CUDA/cuDNN version: NA
- GPU model and memory: NA

**Describe the current behavior**

When invoking `keras.Model.fit` with two `data.Dataset`s, one is for training, and one is validation, so that the training one is infinite with `epochs` and `steps_per_epoch` provided to `fit`, and the validation one is finite without any additional parameters to `fit`, the progress bar shows incorrect number of steps after the first validation. In particular, the number of steps per epoch gets set to the number of steps in the validation set.

**Describe the expected behavior**

The prescribed `steps_per_epoch` remains the same in the progress bar for all epochs.

**Code to reproduce the issue**

See below.

**Other info / logs**

I believe the problem is due to this modification when the input is exhausted:

https://github.com/tensorflow/tensorflow/blob/r1.15/tensorflow/python/keras/engine/training_v2.py#L136

The number of steps gets overwritten. However, `self.params` is shared across both the training and validation callbacks, which essentially misleads the progress bar for training. The state share is potentially due to the following line:

https://github.com/tensorflow/tensorflow/blob/r1.15/tensorflow/python/keras/engine/training_v2.py#L347

In other words, `validation_callbacks` is based on `training_callbacks`. `configure_callbacks` should probably take `callbacks` instead.

**Comment by ravikyram** [comment]
@IvanUkhov 
In order to expedite the trouble-shooting process, please provide a minimal standalone code to reproduce the issue reported here. Thanks!

**Comment by IvanUkhov** [comment]
@ravikyram, it is good that you asked for an example. It turned out it was only about those datasets whose size could not be known in advance. I had to use a generator to reproduce.

```python
import numpy as np
import tensorflow as tf

tf.enable_eager_execution()

def generator():
    for _ in range(100):
        yield [1, 1], 1

training = tf.data.Dataset \
    .from_generator(
        generator=generator,
        output_types=(tf.float64, tf.float64),
        output_shapes=(tf.TensorShape([2]), tf.TensorShape([])),
    ) \
    .batch(2) \
    .repeat()

validation = tf.data.Dataset \
    .from_generator(
        generator=generator,
        output_types=(tf.float64, tf.float64),
        output_shapes=(tf.TensorShape([2]), tf.TensorShape([])),
    ) \
    .batch(2)

model = tf.keras.Sequential([
    tf.keras.layers.InputLayer(2),
    tf.keras.layers.Dense(1),
])
model.compile(loss='mean_squared_error', optimizer='adam')
model.fit(
    x=training,
    validation_data=validation,
    epochs=10,
    steps_per_epoch=20,
)
```

**Comment by ravikyram** [comment]
I have tried on colab with TF version 1.15.0-rc0,1.15.0-rc1 and was able to reproduce the issue.Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/608e6fcff91f3b80d8025f7f9c4876f6/untitled231.ipynb). Thanks!

**Comment by IvanUkhov** [comment]
I have an idea of how it could be fixed, and I will try to explore it in #32847.
